src/desdemona/plugins/swift.clj

```diff
@@ -0,0 +1,127 @@
+  [auth-response]
+  (-> auth-response :access :token))
+
+(defn authenticate
+  "Authenticate with Cloud Files and return the authentication response. This
+  response will include an auth token that can be used for subsequent
+  requests, as well as a service catalog that will indicate the URL to use
+  for, for example, Cloud Files."
+  [auth-url username api-key]
+  (let [body (json/generate-string {"auth" {"RAX-KSKEY:apiKeyCredentials"
+                                            {"username" username
+                                             "apiKey" api-key}}})
+        response (http/post auth-url
+                            {:body body
+                             :content-type :json
+                             :accept :json
+                             :throw-entire-message true
+                             :as :stream})
+        decoded (json/decode-stream (bs/to-reader (response :body))
+                                    ->kebab-case-keyword)]
+    decoded))
+
+(defn create-container
+  "Create a container on Cloud Files. This requires an auth token, which is
+  passed as a request header. If the container already exists, nothing
+  happens."
+  [cf-url auth-token container-name]
+  (let [url (str cf-url "/" container-name)
+        response (http/put url
+                           {:accept :json
+                            :throw-entire-message true
+                            :headers {"X-Auth-Token" auth-token}})]
+    (:status response)))
+
+(defn write-file
+  "Encode the given contents to JSON and write them to the given file on
+  Cloud Files, at the given URL. This requires an auth token, which is passed
+  as a request header. If the file already exists, it is overwritten."
+  [cf-url auth-token container-name file-name contents]
+  (let [url (str cf-url "/" container-name "/" file-name)
+        body (json/generate-string contents)
+        response (http/put url
+                           {:body body
+                            :content-type :json
+                            :throw-entire-message true
+                            :headers {"X-Auth-Token" auth-token}})]
+    (:status response)))
+
+(defn calculate-container-name
+  "We create one container per day, named based on the date. This function
+  generates a container name indicating today's date."
+  []
+  (f/unparse (f/formatter "yyyy-MM-dd") (t/now)))
+
+(defn calculate-file-name
+  "We create one file per batch, named based on the current time. The
+  resolution is intended to be small enough such that each filename will be
+  unique per batch."
+  []
+  (f/unparse (f/formatter "HH:mm:ss-SSS") (t/now)))
+
+(defrecord SwiftWriteRows
+           [auth-url username api-key]
+  p-ext/Pipeline
+
+  (read-batch
+    [_ event]
+    (function/read-batch event))
+
+  (write-batch
+    [_ {:keys [onyx.core/results]}]
+    (let [segments (map :message (mapcat :leaves (:tree results)))]
+      (when (not-empty segments)
```

Comment:
(also as a general rule don't use empty/not-empty when simple nil/not-nil will do, e.g. you should almost always use `(seq x)` in favor of `(not (empty? x))` or `(not-empty x)`)